src/charm.py

```diff
@@ -138,6 +141,63 @@ def __init__(self, *args):
         self.framework.observe(self.on.config_changed, self._configure)
         self.framework.observe(self.on[PEER_RELATION_NAME].relation_created, self._configure)
         self.framework.observe(self.on[PEER_RELATION_NAME].relation_changed, self._configure)
+        self.framework.observe(self.on.authorize_charm_action, self._on_authorize_charm_action)
+
+    def _on_authorize_charm_action(self, event: ActionEvent):
+        """Authorize the charm to interact with Vault."""
+        if not self.unit.is_leader():
+            event.fail("This action can only be run by the leader unit")
+            return
+        logger.info("Authorizing the charm to interact with Vault")
+        if not self._api_address:
+            event.fail("API address is not available. Something is wrong.")
+            return
+        token = event.params["token"]
+        vault = Vault(
+            url=self._api_address, ca_cert_path=self.tls.get_tls_file_path_in_charm(File.CA)
+        )
+        vault.set_token(token)
+        vault.enable_audit_device(device_type="file", path="stdout")
+        vault.enable_approle_auth()
+        vault.configure_charm_access_policy(name=CHARM_POLICY_NAME)
+        # TODO: What should this actually be set to?
+        cidrs = [f"{self._bind_address}/24"]
+        role_id = vault.configure_approle(
+            name="charm",
+            cidrs=cidrs,
+            policies=[CHARM_POLICY_NAME],
+        )
+        vault_secret_id = vault.generate_role_secret_id(name="charm", cidrs=cidrs)
+        self._create_approle_secret(role_id, vault_secret_id)
+        self.on.config_changed.emit()
```

Comment:
Why call the event instead of the handler (`self._configure`)?

Reply:
To put another way, it feel safer to think about the event (the config changed) than what needs to happen because of that event (i.e. these functions need to be called, in this order, at this time) and let the framework decide how to react.

Reply:
After thinking about this some more, I think you're right that it's best to call `_configure` directly because in this case, the config hasn't changed (from a juju perspective), we're just asking the charm to reconfigure itself based on something we've done.

I'll make this change.